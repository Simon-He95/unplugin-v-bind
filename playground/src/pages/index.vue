<script setup lang="ts">
const color = ref('red')
const move = ref(10)
const add = () => move.value += 10
</script>

<template>
  <div class="color">
    color
  </div>
  <div class="move" @click="add">
    move
  </div>
</template>

<style scoped>
.color {
  color: v-bind(color)
}

.move {
  transform: translateX(v-bind(move)px);
  width: v-bind(move) * 10px;
  height: calc(v-bind(move) * 10px)
}
</style>
