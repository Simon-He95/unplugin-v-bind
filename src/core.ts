const styleReg = /<style[^>]+>(.*)<\/style>/s
const vbindReg = /(calc)?[\w\(]*(v-bind\([^)]+\))([^);}:]+)/g

export function vbindTransform(code: string) {
  const match = code.match(styleReg)
  if (!match)
    return code
  const _class = match[1]
  // 匹配是否有v-bind()
  for (const item of _class.matchAll(vbindReg)) {
    const [origin, jump = '', vbind, computed] = item
    if (jump)
      continue
    const calc = computed.trim().replace(/\s+/, ' ').split(' ').filter(Boolean)
    if (!calc.length)
      continue
    const _calc = calc.length === 1
      ? `calc(${vbind} * 1${calc[0]})`
      : `calc(${vbind} ${calc.join(' ')})`
    const replacer = jump + vbind + computed
    code = code.replace(origin, origin.replace(replacer, _calc))
  }
  return code
}
