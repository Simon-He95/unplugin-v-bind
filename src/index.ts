export * from './types'
export * from './unplugin'
