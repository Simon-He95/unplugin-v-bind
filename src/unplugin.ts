import { createUnplugin } from 'unplugin'
import { createFilter } from '@rollup/pluginutils'
import { vbindTransform } from './core'
import type { Options } from './types'

const unplugin = createUnplugin((options: Options = {}): any => {
  const filter = createFilter(options.include, options.exclude)
  return [
    {
      name: 'unplugin-v-bind',
      enforce: 'pre',
      transformInclude(id: string) {
        return filter(id)
      },
      async transform(code: string, id: string) {
        // 只处理.vue文件
        if (!id.endsWith('.vue'))
          return code
        return vbindTransform(code)
      },
      handleHotUpdate({ file, server }: any) {
        if (file.endsWith('.vue')) {
          console.log('reloading vue file...')
          server.ws.send({
            type: 'full-reload',
            path: '*',
          })
        }

        return []
      },
    },
  ]
})

export const vitePlugin = unplugin.vite
export const rollupPlugin = unplugin.rollup
export const webpackPlugin = unplugin.webpack
export const esbuildPlugin = unplugin.esbuild
