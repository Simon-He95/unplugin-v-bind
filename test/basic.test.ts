import fsp from 'fs/promises'
import { describe, expect, it } from 'vitest'
import { vbindTransform } from '../src/core'

describe('vbindTransform Test', async () => {
  const dir = await fsp.readdir('./src/vue')
  await Promise.all(dir.map(async (item) => {
    const code = await fsp.readFile(`./src/vue/${item}`, 'utf-8')
    it(`${item}`, async () => {
      expect(vbindTransform(code)).toMatchSnapshot()
    })
  }))
})
