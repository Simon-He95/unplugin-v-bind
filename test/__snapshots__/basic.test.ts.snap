// Vitest Snapshot v1

exports[`vbindTransform Test > 1.vue 1`] = `
"<script setup lang=\\"ts\\">
const color = ref('red')
const move = ref(10)
const add = () => move.value += 10
</script>

<template>
  <div class=\\"color\\">
    color
  </div>
  <div class=\\"move\\" @click=\\"add\\">
    move
  </div>
</template>

<style scoped>
.color {
  color: v-bind(color)
}

.move {
  transform: translateX(calc(v-bind(move) * 1px));
  width: calc(v-bind(move) * 10px);
  height: calc(v-bind(move) * 10px)
}
</style>
"
`;
